### ISSUE TITLE
OXC emitDecoratorMetadata: divergences from tsc

### ISSUE THREAD

**Issue by kylecannon** [oxc-project/oxc#21922]
# OXC `emitDecoratorMetadata`: 7 divergences from `tsc`

Empirical comparison of `tsc` vs OXC (via Rolldown) for `experimentalDecorators` + `emitDecoratorMetadata`. Each case is a `design:type` (or runtime behaviour) divergence from the TypeScript compiler spec.

## Environment (all bugs)

- `rolldown` 1.0.0-rc.16 with `transform.decorator.{ legacy: true, emitDecoratorMetadata: true }`
- `@oxc-project/runtime` 0.128.0
- `typescript` 5.9.3 (baseline)
- Node.js v22.22.0, macOS (Darwin 24.6.0)

---

## Bug 1 — `type` alias to a primitive emits `Object`

```ts
import 'reflect-metadata';
function D(): PropertyDecorator { return () => {}; }

type Uuid = string;
type Count = number;
type IsActive = boolean;

class Entity {
  @D() id!: Uuid;
  @D() count!: Count;
  @D() active!: IsActive;
}

const e = new Entity();
for (const f of ['id', 'count', 'active']) {
  const t = Reflect.getMetadata('design:type', e, f);
  console.log(`${f}: ${t?.name ?? t}`);
}
```

**Expected (tsc):**
```
id: String
count: Number
active: Boolean
```

**Actual (OXC):**
```
id: Object
count: Object
active: Object
```

OXC bakes the literal `Object` into each `__decorateMetadata` call — alias resolution is skipped.

**Suggested fix area:** When OXC's legacy decorator metadata transform encounters a `TSTypeReference` whose name resolves to a `TSTypeAliasDeclaration` with a primitive keyword body (`string`, `number`, `boolean`, `bigint`), emit the corresponding constructor (`String`, `Number`, `Boolean`, `BigInt`). `tsc` handles this via its checker (`serializeTypeReferenceNode` → `resolver.getTypeReferenceSerializationKind` in `src/compiler/transformers/typeSerializer.ts`). OXC would need a name-based alias-resolution pass.

**Severity:** Libraries that consume `design:type` (NestJS Swagger, TypeORM, `@automapper/classes`, `class-transformer`) receive `Object` instead of the intended primitive. Consumers that discard `Object` silently drop fields whose annotation is an alias.

---

## Bug 2 — Intersection of a primitive and a brand emits `Object`

A "branded primitive" type — `string & { readonly __brand: 'UserId' }` — is a common nominal-typing pattern. `tsc` resolves the intersection to its primitive constituent and emits `String` / `Number`. OXC emits `Object`.

Note: when the intersection contains a real structural object member (e.g. `string & { x: number }`), `tsc` itself emits `Object`. The bug is specifically about brand-only intersections, where the brand object contributes only a phantom symbol property. See [microsoft/TypeScript#41201](https://github.com/microsoft/TypeScript/issues/41201) for `tsc`'s working-as-intended classification.

```ts
import 'reflect-metadata';
function D(): PropertyDecorator { return () => {}; }

type UserId = string & { readonly __brand: 'UserId' };
type Count = number & { readonly __brand: 'Count' };

class Entity {
  @D() userId!: UserId;                            // tsc: String,  OXC: Object
  @D() count!: Count;                              // tsc: Number,  OXC: Object
  @D() plainStringBrand!: string & { x: number }; // tsc: Object,  OXC: Object — agrees, not the bug
}

const e = new Entity();
for (const f of ['userId', 'count', 'plainStringBrand']) {
  const t = Reflect.getMetadata('design:type', e, f);
  console.log(`${f}: ${t?.name ?? t}`);
}
```

**Expected (tsc):**
```
userId: String
count: Number
plainStringBrand: Object
```

**Actual (OXC):**
```
userId: Object
count: Object
plainStringBrand: Object
```

**Suggested fix area:** When OXC encounters a `TSIntersectionType`, walk the constituents. If exactly one is a primitive keyword type while the rest are interface/object types whose only members are phantom branding properties, emit the primitive constructor. A simpler heuristic — "if any constituent is a primitive keyword, emit that primitive" — would cover the dominant branded-primitive case, though it would also match `string & { x: number }` (which `tsc` intentionally classifies as `Object`).

**Severity:** Branded primitives (`UserId`, `OrderId`, `Email`, `Cents`, …) are a widespread pattern. Under OXC every such field loses its primitive-typed runtime metadata and downstream consumers (Swagger schema, class-validator, AutoMapper, ORM column types) silently degrade.

---

## Bug 3 — `ReadonlyArray<T>` emits `Object` (short form `readonly T[]` works)

```ts
import 'reflect-metadata';
function D(): PropertyDecorator { return () => {}; }

class Entity {
  @D() shortForm!: readonly string[];                          // tsc: Array,  OXC: Array  — OK
  @D() genericForm!: ReadonlyArray<string>;                    // tsc: Array,  OXC: Object — DIVERGES
  @D() nestedReadonly!: ReadonlyArray<ReadonlyArray<number>>;  // tsc: Array,  OXC: Object — DIVERGES
}

const e = new Entity();
for (const f of ['shortForm', 'genericForm', 'nestedReadonly']) {
  const t = Reflect.getMetadata('design:type', e, f);
  console.log(`${f}: ${t?.name ?? t}`);
}
```

**Expected (tsc):**
```
shortForm: Array
genericForm: Array
nestedReadonly: Array
```

**Actual (OXC):**
```
shortForm: Array
genericForm: Object
nestedReadonly: Object
```

OXC emits `typeof ReadonlyArray === "undefined" ? Object : ReadonlyArray` for the generic form. `ReadonlyArray` is a TS-only utility type with no runtime binding, so the guard always picks the `Object` branch.

**Suggested fix area:** Special-case the well-known TS-only utility types whose runtime equivalent is a built-in constructor: `ReadonlyArray<T>` → `Array`, `ReadonlyMap<K, V>` → `Map`, `ReadonlySet<T>` → `Set`. `tsc` handles these via the type checker; OXC needs a name-based recognition pass at transform time.

**Severity:** Affects any `ReadonlyArray<T>` (and likely `ReadonlyMap` / `ReadonlySet`) used as a decorated property type. Consumers like NestJS Swagger schema, class-validator, and `@automapper/classes` lose array-shape metadata for these fields.

---

## Bug 4 — `strictNullChecks: false` ignored; `T | null` and `prop?: T` emit `Object` regardless of tsconfig

TypeScript's decorator metadata depends on `strictNullChecks`:

- With `strictNullChecks: true` (default): `T | null` → `Object` (working as intended, per [microsoft/TypeScript#12703](https://github.com/microsoft/TypeScript/issues/12703)).
- With `strictNullChecks: false`: `null` and `undefined` are elided from unions, so `T | null` → `T` and `prop?: T` → `T`.

OXC always behaves as if `strictNullChecks` is on. Even when an explicit `tsconfig.json` sets `"strict": false, "strictNullChecks": false`, OXC still emits `Object` for `T | null` and for optional properties (`prop?: T`), because optional properties implicitly expand to `T | undefined`.

```ts
// tsconfig: { "strict": false, "strictNullChecks": false }
import 'reflect-metadata';
function D(): PropertyDecorator { return () => {}; }

class Entity {
  @D() id: string;
  @D() name: string | null;
  @D() count: number | null;
  @D() label?: string;
  @D() value?: number;
}

const e = new Entity();
for (const f of ['id', 'name', 'count', 'label', 'value']) {
  const t = Reflect.getMetadata('design:type', e, f);
  console.log(`${f}: ${t?.name ?? t}`);
}
```

**Expected (tsc with `strictNullChecks: false`):**
```
id: String
name: String
count: Number
label: String
value: Number
```

**Actual (OXC with the same tsconfig):**
```
id: String
name: Object
count: Object
label: Object
value: Object
```

Both a plain `transform.decorator` invocation and an invocation alongside an explicit tsconfig with `"strictNullChecks": false` produce the same wrong output — the option is being ignored.

**Suggested fix area:** When OXC's legacy decorator metadata transform processes a union type, consult `compilerOptions.strictNullChecks` (and the `strict` aggregate). If strict-null-checks is disabled, drop `null` and `undefined` constituents before serializing — matching `tsc`'s `serializeUnionOrIntersectionConstituents`. This covers both explicit `T | null` unions and optional-property shorthand (`prop?: T`, which expands to `T | undefined`). The transform option is reachable via the rolldown plugin config; a tsconfig consult would also need a hook for the standalone `oxc_transformer` crate API.

**Severity:** Low (config-driven, non-default tsconfig). Codebases predating strict mode commonly rely on `T | null` and `prop?: T` resolving to `T`; under OXC every such field becomes `Object`.

---

## Bug 5 — `typeof X === "function"` guard falls through to `Object` for cross-file enums

When OXC cannot statically prove a type reference is a class constructor (the dominant case: a cross-file imported enum), it emits a runtime guard:

```js
typeof (_ref = typeof X !== "undefined" && X) === "function" ? _ref : Object
```

This guard is correct for class constructors (which are functions) but **always falls through to `Object` for enum objects**, because TypeScript enums lower to plain objects (`{ A: "a", B: "b" }`) — `typeof` is `"object"`, not `"function"`.

Cross-file imports are essential for triggering this: same-file enums are inlined by OXC and avoid the guard. Only the cross-file shape exhibits the bug.

**`enums.ts`:**
```ts
export enum StringEnum { A = 'a', B = 'b' }
export enum NumericEnum { X = 1, Y = 2 }
```

**`source.ts`:**
```ts
import 'reflect-metadata';
import { StringEnum, NumericEnum } from './enums';

function D(): PropertyDecorator { return () => {}; }

class Source {
  @D() str!: StringEnum;   // tsc: String, OXC: Object
  @D() num!: NumericEnum;  // tsc: Number, OXC: Object
}

const inst = new Source();
for (const f of ['str', 'num']) {
  const t = Reflect.getMetadata('design:type', inst, f);
  console.log(`${f}: ${t?.name ?? String(t)}`);
}
```

**Expected (tsc):**
```
str: String
num: Number
```

**Actual (OXC):**
```
str: Object
num: Object
```

The transformed bundle shows the guard literally:
```js
__decorateMetadata(
  "design:type",
  typeof (_ref = typeof StringEnum !== "undefined" && StringEnum) === "function" ? _ref : Object
)
```

At runtime `StringEnum` is `{ A: "a", B: "b" }` — `typeof === "object"` — so the guard picks `Object`.

**Suggested fix area:** A `typeof X === "function"` runtime test cannot distinguish "this enum object" from "this binding is undefined." The fix needs to happen at transform time: for `TSEnumDeclaration` references, classify the enum's member-initializer kinds at the import site (string-only → `String`; numeric / mixed → `Number`), mirroring `tsc`'s `getTypeReferenceSerializationKind` logic for enums.

**Severity:** Cross-file enums are extremely common in real codebases. Under OXC every such decorated field loses its enum metadata: `@automapper/classes` discards the field, NestJS Swagger schema becomes `{}`, class-validator and TypeORM/TypedORM column-type detection degrade to permissive defaults.

---

## Bug 6 — Forward / cyclic class references silently emit `Object` instead of throwing TDZ

Same `typeof X !== "undefined" && X` guard as Bug 5, different trigger. For a class defined textually later in the same module (or imported through a circular module graph), the binding has been hoisted (`var X`) but is still `undefined` at the moment `__decorate` runs. The guard short-circuits and `design:type` is silently recorded as `Object`.

This is a **silent-failure regression vs `tsc`**. `tsc` emits a bare identifier reference and relies on the temporal dead zone (TDZ) to throw `ReferenceError: Cannot access 'X' before initialization` — a loud, debuggable signal. OXC's guard masks that signal entirely.

```ts
import 'reflect-metadata';
function D(): PropertyDecorator { return () => {}; }

class Source {
  @D() laterRef!: LaterClass;   // forward reference
}

class LaterClass { tag = 'later'; }

const inst = new Source();
const t = Reflect.getMetadata('design:type', inst, 'laterRef');
console.log(`laterRef: ${t?.name ?? String(t)}`);
void LaterClass;
```

**Expected (tsc):**
```
ReferenceError: Cannot access 'LaterClass' before initialization
```

**Actual (OXC):**
```
laterRef: Object
```

OXC emits `var LaterClass` (hoisted, `undefined`). The guard's `typeof !== "undefined"` check is false, `_ref` becomes `false`, and `Object` is recorded silently.

**Why this is worse than the other bugs:**

- The transformed AST is internally consistent. The bug only manifests at run time, depending on when `__decorate` runs relative to the binding's actual assignment.
- Downstream consumers see `Object` and behave as if the field was simply not annotated. There is no error, no log, no signal.
- The same guard propagates to circular-import scenarios under rolldown's module-init ordering. With a circular `import` graph, a cross-file class reference can hit the same `undefined`-at-decoration-time window.

**Suggested fix area:** Drop the `typeof X !== "undefined"` half of the guard and emit just the bare identifier (or `typeof X === "function" ? X : Object` without the existence check). A forward / TDZ reference would then throw `ReferenceError`, matching `tsc`. The guard exists to prevent `ReferenceError` — but the cure (silent `Object`) is strictly worse: the developer has no observability into the silent case, whereas the loud error points them at the exact line.

This may need to be filed against rolldown rather than the OXC core transformer if the guard is emitted by the rolldown integration layer rather than the standalone `oxc_transformer` crate.

**Severity:** TypeORM / TypedORM forward-referenced entity classes degrade to `Object` (framework reports a misleading downstream error). NestJS DI cyclic-imported provider classes can record `Object` as their constructor-param type, causing silent DI mismatch. `@automapper/classes` drops the field.

---

## Bug 7 — `accessor` keyword properties always emit `design:type` as `Object`

When a decorated class member uses the TypeScript 4.9+ `accessor` keyword (auto-accessor syntax), OXC's legacy decorator metadata transform never invokes the type serializer. Every `@D() accessor x: T` records `design:type` as `Object`, regardless of `T`.

This affects all annotation kinds — primitives, classes, enums, arrays, `Promise`, `Date`, etc. — even those that OXC handles correctly at a plain `PropertyDefinition`.

```ts
import 'reflect-metadata';

function D(): any { return () => {}; }
const DA: any = D;   // widened to any — see note below

enum Color { Red = 'red', Blue = 'blue' }
class Addr { street: string = ''; }

class Entity {
  @DA() accessor name: string = '';                               // tsc: String,   OXC: Object
  @DA() accessor count: number = 0;                               // tsc: Number,   OXC: Object
  @DA() accessor color: Color = Color.Red;                        // tsc: String,   OXC: Object
  @DA() accessor addr: Addr = new Addr();                         // tsc: Addr,     OXC: Object
  @DA() accessor pending: Promise<string> = Promise.resolve(''); // tsc: Promise,  OXC: Object
}

const e = new Entity();
for (const f of ['name', 'count', 'color', 'addr', 'pending']) {
  const t = Reflect.getMetadata('design:type', e, f);
  console.log(`${f}: ${t?.name ?? t}`);
}
```

Note: `const DA: any = D` is required because OXC's legacy transform rejects a function-typed decorator on an `accessor` member. Widening to `any` is a workaround for that separate issue. The `Object` fallthrough in `design:type` happens regardless of how the decorator is typed.

**Expected (tsc):**
```
name: String
count: Number
color: String
addr: Addr
pending: Promise
```

**Actual (OXC):**
```
name: Object
count: Object
color: Object
addr: Object
pending: Object
```

OXC emits the literal `Object` baked into every call. The constructor name (`String`, `Number`, `Addr`, etc.) does not appear anywhere in the emitted output for any of the five fields.

**Hypothesis:** The dispatch in OXC's metadata serializer handles `PropertyDefinition` and `MethodDefinition` but does not handle `AccessorProperty`, so it falls through to the default emit path which records `Object`.

**Suggested fix area:** Treat `AccessorProperty` like `PropertyDefinition` for `design:type` serialization. The runtime semantics differ (auto-accessors install a getter/setter pair backed by a private field), but the metadata emit should be identical — the TypeScript spec for `emitDecoratorMetadata` makes no exception for accessor members.

**Severity:** The `accessor` keyword (TS 4.9, Nov 2022) is increasingly used in DI-heavy NestJS code, libraries adopting stage-3 decorators while keeping legacy metadata for back-compat, and ORM models. For any such code under OXC, every decorated `accessor` field loses its `design:type`.

Empirically: across 33 testable accessor patterns in our test matrix, `tsc` emits a concrete type and OXC emits `Object` in 17 cells. The remaining 16 are cells where `tsc` itself emits `Object` (genuine unions, generics, etc.). There is no accessor cell where OXC emits a non-`Object` type — the serializer is unconditionally skipped.

Divergent cells:

| Pattern | tsc `design:type` | OXC `design:type` |
| --- | --- | --- |
| `string` primitive | `String` | `Object` |
| `type` alias to `string` | `String` | `Object` |
| Branded primitive (`string & brand`) | `String` | `Object` |
| String enum | `String` | `Object` |
| Numeric enum | `Number` | `Object` |
| `T[]` short array | `Array` | `Object` |
| `Array<T>` | `Array` | `Object` |
| `ReadonlyArray<T>` | `Array` | `Object` |
| Tuple | `Array` | `Object` |
| `Promise<T>` | `Promise` | `Object` |
| `Date` (builtin) | `Date` | `Object` |
| Function type | `Function` | `Object` |
| `bigint` primitive | `BigInt` | `Object` |
| Optional-marker class | the class | `Object` |
| Template literal type | `String` | `Object` |
| `accessor` keyword (same type) | `String` | `Object` |
| Computed property key | `String` | `Object` |
| Cross-file imported enum | `String` | `Object` |

**Comment by Dunqing** [comment]
Thanks for the detailed writeup — the per-cell tables made triage straightforward.

#7 is fixed in #21966.

For #1, #2, #3, #5: these all reduce to resolving a type reference back to its declaration, which is a type-checker job. The transformer doesn't run one, and `babel-plugin-transform-typescript-metadata` has the same constraint and chose not to add a name-based heuristic either. We're inclined to match Babel here.

#4 is interesting — worth noting that the Babel plugin doesn't actually read `strictNullChecks`. It unconditionally elides `null` and `undefined` from unions regardless of tsconfig. That technically diverges from tsc when `strictNullChecks: true` (where `T | null` → `Object`), but it gives the metadata that NestJS / class-validator / TypeORM actually want, and avoids any tsconfig plumbing. We'd consider the same shape here.

#6 is more of a tradeoff than a clear bug. The `typeof X !== "undefined"` guard prevents TDZ throws on forward or circular class references; without it, silent `Object` becomes a runtime crash. Happy to discuss alternatives — would you mind opening this as its own issue?

**Comment by kylecannon** [comment]
@Dunqing 

On #1, #2, #3: agree with the closes, those are the right call. On the rest:

For #4, strong agree on the babel-style approach. tsc's `T | null` to `Object` under `strictNullChecks: true` is widely viewed as annoying rather than useful by exactly the consumers that read this metadata (class-validator, NestJS Swagger, TypeORM, AutoMapper). Unconditional elision matches what real downstream code expects and avoids tsconfig plumbing. Happy to see this shape land.

For #5, want to flag this one again, narrowly, because I think it sits in a different bucket than #1, #2, #3.

The issue is not type-reference resolution. It is the structure of the emitted runtime guard. Concretely, OXC emits:

```js
typeof (_ref = typeof X !== "undefined" && X) === "function" ? _ref : Object
```

This guard is a guaranteed fallthrough to `Object` for every enum, because TypeScript enums lower to plain objects (`{ A: 'a', B: 'b' }`) where `typeof === "object"`, never `"function"`. So for any decorated property typed as a cross-file imported enum, the runtime metadata is unconditionally `Object`.

Babel does not have this problem because it emits the binding identifier directly with no guard at all. `Reflect.getMetadata('design:type', ...)` returns the enum object itself, which NestJS Swagger and similar consumers can introspect via `Object.values(t)` to recover member names. So in this specific case, "match babel" is not a no-op closure. It would mean dropping the `typeof === "function"` test (or at least broadening it to also accept `"object"`-shaped non-undefined bindings).

I do want to acknowledge what I think you are protecting against here. The cross-file imported case genuinely cannot be classified as enum-vs-class without cross-module resolution, and OXC's transformer is per-file. So a fully type-aware fix is not cheap. But the current emit is strictly worse than babel rather than equivalent: babel produces a useful object, OXC produces `Object` for every cross-file enum. Could we discuss this on its own thread rather than closing it under the type-resolution rubric? Even just acknowledging the structural distinction in the close note would help.

For #6, will file a fresh issue with a clean repro. Quick preview of the framing so you know what is coming:

The `typeof X !== "undefined"` half of the guard does not prevent a TDZ crash. It silently swallows the TDZ access and substitutes `Object`. That is exactly the `var`-hoisting footgun (use-before-init silently produces `undefined`) that ES2015 introduced TDZ to eliminate. In effect the guard re-creates the hazard the spec was designed to fix.

Concretely:

| Scenario | tsc | babel | OXC |
| --- | --- | --- | --- |
| forward-ref to class in module | ReferenceError | ReferenceError | Object (silent, no error) |

tsc and babel emit a bare identifier and rely on TDZ to throw. The throw points at the offending line and gets resolved in minutes. OXC's silent `Object` propagates through DI containers, ORM column-type detection, AutoMapper field selection, and class-transformer hydration, where it surfaces hours or days later as "field disappeared from the API response" or "wrong provider injected" with no stack trace pointing back to the cause.

The fix looks small from the outside (gate the existence-check on whether the binding has a resolved symbol), so legitimate ambient or `declare`-erased bindings keep the guard while same-module class references go through bare-identifier emit and let TDZ fire. Will write that up properly with the repro and link it back to this thread.

For #7, thanks for the quick fix. Will verify shortly.
